README.md
# ArtSpatter
Art Community Management System by Jake Gustafson

The table models and schemas are in `./models`. Some of them are denormalized
for performance (Drkušić, 2016).

Comments in the code marked "(future)" are not for the 8-week project,
but are being considered for future versions.


## Configuration
(These steps are only necessary if downloaded from git--These pieces are
included in the workspace.)

- Create your config/auth.config.js similar to:
```JavaScript
module.exports = {
    secret: 'this-is-no-secret', // change this in your instance.
};
```
- Create your config/db.config.js similar to:
```JavaScript
module.exports = {
  HOST: 'localhost', // 127.0.0.1
  PORT: '27017',
  DB: 'artspatter',
}
```
- Create your `client/.env` similar to defaults below--PORT default
  comes from "npm start" internally which uses the PORT environment 
  variable (API_URL's port must match API_PORT in `.env`):
```
PORT=3000
API_URL=http://localhost:5000
```
- Create your `.env` similar to the defaults below (CLIENT_ORIGIN must
  match PORT in client/.env):
```
API_PORT = 5000
CLIENT_ORIGIN = http://localhost:3000
```


## Development
For how to maintain the code, see contributing.md.


## Citation Style

Some of the references are cited in the code as opposed to this page.
Using markdown format, `*` denotes italics where specified by APA 7,
since this file and the code are plain text files (though this file
will change the markings to italics if viewed in a markdown viewer or
if converted to another format such as HTML or PDF).


## References
- BezKoder. (2019a, October 19). Node.js + MongoDB: User authentication & authorization with JWT. *BezKoder*. https://bezkoder.com/node-js-mongodb-auth-jwt/
- BezKoder. (2019b, October 19). React JWT authentication (Without Redux) example. *BezKoder*. https://bezkoder.com/react-jwt-auth/
- Drkušić, E. (2016, March 17). *Denormalization: When, why, and how*. Vertabelo Data Modeler. https://vertabelo.com/blog/denormalization-when-why-and-how/
- *Mongoose relationships tutorial*. (n.d.). Vegibit. https://vegibit.com/mongoose-relationships-tutorial/
- *Mongoose v5.10.15: SchemaTypes*. (n.d.). Mongoose. https://mongoosejs.com/docs/schematypes.html


changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).


## [git] - 2021-07-26
### Changed
- Change the default API_PORT (set in .env) from 8085 (formerly 5000) to 56765.
- Change the client origin from http://localhost:3000 to http://localhost:54445


## [git] - 2020-12-23
### Added
- (readme and contributing) Document the process of installing and
  running.

### Changed
- Switch to yarn.
  - Improve related documentation.

### Removed
- Remove npm-related documentation from contributing and readme files.


## [git] - 2020-12-22
### Changed
- Release 0.9.0


## [git] - 2020-12-13
### Added
- store c time (creation Date)


## [git] - 2020-12-12
### Added
- `node-thumbnail` to add thumbnails (Only save posts
  when the file is a real image, according to whether
  the thumbnail operation reports success).

### Fixed
- `fs` and `mv` to save uploads properly.
- Install filesystem and thumbnailing libraries using npm.
- Move processable images to the public directory and generate
  thumbnails upon upload, saving paths to the database.


## [git] - 2020-12-11
### Fixed
- Refactor html: Migrate most or all to Tailwind and semantic HTML.
- Fix use of multer for uploads.
- Install and implement `tailwind-navbar-react`


## [git] - 2020-12-11
### Added
- Install the react-infinite-scroll-component using npm.

### Changed
- Change code to semantic HTML.

### Fixed
- Install tailwind-navbar-react using npm.
- Implement it to redo the website’s nav for responsiveness.
- refactor html, migrate most to Tailwind and semantic HTML


## [git] - 2020-12-10
### Added
Create a logo.

### Changed
- Split the nav code into a separate component.

### Fixed
- Make a link load the upload form.
- Change className in several cases to use Tailwind.


## [git] - 2020-12-09
### Added
- Install multer for processing image uploads.
- Code a new route and new API feature that saves the image then adds
  metadata for it in the database.
- Create a new upload form that sends data to the new backend route.


## [git] - 2020-12-06
### Added
- placeholder for post system
- tests
- Create a PrivacySchema to determine on which profile fields the user
  can change the privacy level (not yet added to git).
- Script the loading of initial data necessary to operate the site and
  to test the functionality, utilizing all models except Vote.

### Changed
- No longer hard-code settings (See readme).
- Refine posting, voting, VoteType, flagging, FlagType, section,
  category, role, and user models (not yet added to git).


## [git] - 2020-12-05
### Added
- (backend) Implement numerical id using node-sequence.
- (frontend) errorLeaf function (and reporting module)
  - encapsulate error message formatting code.
- (backend) Create scripts to clear the database.
- (backend) Create functions to load initial database entries to make site usable.

### Fixed
- (frontend) Fix error reporting (React will not accept an error object
  as a state value--use new errorLeaf function).


## [git] - 2020-11-28
### Added
- JWT Authentication

### Changed
- Rename index.js to server.js.


## [git] - 2020-11-27
### Added
- eslint

### Changed
- Rename jsx files to use jsx extension (helps with linting).
- Switch to airbnb coding style.

### Fixed
- Remove duplicate dependency in react (resolves issue #13).
- (quality.py) Properly handle both stdout and stderr.

### Removed
- jshint


## [git] - 2020-11-25
### Added
- quality.py
- Add placeholders for per-category and per-role priveleges to roles:
  Allow CRUD granularly (C, R, U, or D separately) for posts in given
  categories and users in given roles.

### Changed
- (quality.py run_command) detect string or array, `decode()` bytestring
- (quality.py run_command) accept a file stream for screen+file output
- (quality.py run_command) fix infinite loop: check output&return
  separately
- Rename foreign keys and primary keys with same names.
- Rearrange indexes and keys for performance.


## [unreleased] - 2020-11-24
### Added
- `friends` to user table


## [git] - 2020-11-21
### Added
- `value` for Ballot (how much the ballot promotes or demotes the post)

### Changed
- Rename BallotName to caption.
- Rename Marking to FlagType.
- Rename Ballot to VoteType.
- Change `FlagUserMarking` and `VoteUserID` to each use a compound index
  (the fields are no longer necessary).
- Make the case same for the model name and the model constant
  (Mongoose Query Population v3.3.1, n.d.).

### Fixed
- Make Mongoose model names lowercase.
- Name the const correctly for Post.

### Removed
- Rename `BallotID` to `n`


## [git] - 2020-11-19
### Changed
- Rename  Ballot ~~Reaction~~


## [git] - 2020-11-17
### Changed
- Rename "Rating" to "Flag"
- Rename "Level" to "Marking"
- Rename "RoleName" to "title"


## [git] - 2020-11-16
### Added
- Finish the DigitalOcean [Getting Started with the MERN
  Stack](https://www.digitalocean.com/community/tutorials/getting-started-with-the-mern-stack)
  tutorial (add custom React components "client/src/App.js" loads them
  via the `./components/Todo` component).


## [git] - 2020-11-11
### Fixed
- Change "yarn" to "npm" in the concurrently command in package.json.
  RESOLVES: Yarn is installed, but running it through npx produces an
  error: See `screenshots/2020-11-11_npx_yarn--error.txt`.
- Change "DB = ‘mongodb://<USER\>:<PASSWORD\>@ds039950.mlab.com:39950/todo’"
  in the `.env` file.
  RESOLVES: MongoNetworkError: failed to connect to server (See
  `2020-11-11_mongod_error.txt`)

## References
- *Mongoose Query Population v3.3.1*. (n.d.). https://mongoosejs.com/docs/3.3.x/docs/populate.html


server.js
const express = require('express');
const bodyParser = require('body-parser');
const cors = require('cors');
// const mongoose = require('mongoose');
const path = require('path');
// const routes = require('./routes/api');
var bcrypt = require("bcryptjs");

var dbConfig = null;

try {
  dbConfig = require('./config/db.config');
}
catch (e) {
  console.error(e);
  console.error();
  console.error('You must first have a proper ./config/db.config.js file. See artspatter/config/db.config.example.js');
  console.error()
  console.error();
  console.error("(ignore errors below)")
  process.exit(1);
}
require('dotenv').config();

const app = express();

var corsOptions = {
  origin: process.env.CLIENT_ORIGIN || 'http://localhost:54445',
};
app.use(cors(corsOptions));
// (BezKoder, 2019a)

const port = process.env.API_PORT || 56765;

app.use(bodyParser.json());

// parse requests of content-type - application/x-www-form-urlencoded
// as per BezKoder (2019a).
app.use(bodyParser.urlencoded({ extended: true }));

const db = require('./models');
const PrivacyLevel = db.privacylevel;
const Role = db.role;
const FlagType = db.flagtype;
const Section = db.section;
const User = db.user;
const Category = db.category;
const VoteType = db.votetype;
const Post = db.post;

// connect to the database
db.mongoose
  .connect(`mongodb://${dbConfig.HOST}:${dbConfig.PORT}/${dbConfig.DB}`, {
    useNewUrlParser: true,
    useUnifiedTopology: true, // (BezKoder, 2019a)
    useFindAndModify: false, // fix DeprecationWarning
    useCreateIndex: true,  // fix DeprecationWarning
  })
  .then(() => {
    console.log('MongoDB is successfully connected.');
    initialPrivacyLevels();
    initialRoles();
    initialFlagTypes();
    initialSections();
    const adminUid = "1"; // The default starting number for node-sequence is 1
    initialUsers(adminUid); // TODO: do the work of initialCategories() ensuring admin user is owner
    initialCategories(adminUid);
    initialTestPosts(adminUid);  // sample content (not required)
    initialVoteTypes();
  })
  .catch((err) => {
    console.error('Connection error', err);
    process.exit(1);
  });


app.get('/', (req, res) => {
  res.json({ message: 'Welcome to an unnamed ArtSpatter instance.' });
});

// routes
require('./routes/auth.routes')(app);
require('./routes/post.routes')(app);
require('./routes/user.routes')(app);

app.use(express.static(__dirname + '/public'));

app.listen(port, () => {
  console.log(`Server running on port ${port}`);
});

const future_notice = "will be displayed on federated sites"
                      + " (tentatively, in future versions with the feature)";

function initialPrivacyLevels() {
  // - number is used in profile (pln is value for privacy fields--see user.model.js)
  // - number must have same meaning for every site (caption may differ)
  PrivacyLevel.estimatedDocumentCount((err, count) => {
    if (!err && count === 0) {
      const entries = [
        [0, "Deleted", null],
        [1, "Archived", null],
        [2, "Private", null],
        [3, "Friends", null],
        [4, "Friends of Friends", null],
        [5, "Logged in to this site", null],
        [6, "Logged in to federation", future_notice],
        [7, "Public", future_notice],
      ];
      for (var index = 0; index < entries.length; index++) {
        var thisE = entries[index];
        let thisV = thisE[1]; // let forces scope to be code block so last value doesn't repeat
        new PrivacyLevel({
          pln: thisE[0],
          caption: thisE[1],
          description: thisE[2],
        }).save((err) => {
          if (err) {
            console.log(`+PrivacyLevel ${thisV} error`, err);
          }
          else {
            console.log(`+PrivacyLevel ${thisV}`);
          }
        });
      }
    }
  });
}

function initialRoles() {
  Role.estimatedDocumentCount((err, count) => {
    if (!err && count === 0) {
      // [active,rid,ccids,rcids,mcids,dcids,rrids,mrids,drids]
      const entries = [
        [true, "user"],
        [true, "moderator"],
        [true, "admin"],
      ];
      for (var index = 0; index < entries.length; index++) {
        var thisE = entries[index];
        let thisV = thisE[1]; // let forces scope to be code block so last value doesn't repeat
        new Role({
          active: thisE[0],
          rid: thisE[1],
        }).save((err) => {
          if (err) {
            console.log(`+Role ${thisV} error`, err);
          }
          else {
            console.log(`+Role ${thisV}`);
          }
        });
      }
    }
  });
}

function initialFlagTypes() {
  FlagType.estimatedDocumentCount((err, count) => {
    if (!err && count === 0) {
      var tldr = "This is only appropriate for mature audiences"
                 + " (may be misleading or provocative).";
      // [active,ftn,caption,description,pln]
      const entries = [
        [true, 0, "General Audiences", "This is appropriate for anyone.", 7],
        [true, 1, "Teen", "This is only appropriate for teens or older (may be shocking).", 5],
        [true, 2, "Mature", tldr, 5],
      ];
      for (var index = 0; index < entries.length; index++) {
        let thisE = entries[index];
        let thisV = thisE[1]; // let forces scope to be code block so last value doesn't repeat
        new FlagType({
          active: thisE[0],
          ftn: thisE[1],
          caption: thisE[2],
          description: thisE[3],
          pln: thisE[4],
        }).save((err) => {
          if (err) {
            console.log(`+FlagType ${thisV}, ${thisE[2]} error`, err);
          }
          else {
            console.log(`+FlagType ${thisV}, ${thisE[2]}`);
          }
        });
      }

    }
  });
}

function initialSections() {
  Section.estimatedDocumentCount((err, count) => {
    if (!err && count === 0) {
      // [active, sid, caption, pln]
      const entries = [
        [true, "art", "Art", 7],
        [true, "forum", "Discussions", 7],
      ];
      for (var index = 0; index < entries.length; index++) {
        var thisE = entries[index];
        let thisV = thisE[1]; // let forces scope to be code block so last value doesn't repeat
        new Section({
          active: thisE[0],
          sid: thisE[1],
          caption: thisE[2],
          pln: thisE[3],
        }).save((err) => {
          if (err) {
            console.log(`+Section ${thisV} error`, err);
          }
          else {
            console.log(`+Section ${thisV}`);
          }
        });
      }
    }
  });
}

function initialUsers(uid) {
  var error = null;
  let adminUid = uid;
  User.estimatedDocumentCount((err, count) => {
    if (!err && count === 0) {
      // [active,uid,username,email,pln,showFtns,display,password,overview]
      const entries = [
        [true, "1", "admin", null,  2, [0,1,2], "Admin", "password", "[The system generated this user.]"],
      ];
      for (var index = 0; index < entries.length; index++) {
        var thisE = entries[index];
        let thisV = thisE[1]; // let forces scope to be code block so last value doesn't repeat
        let passwordPlainText = thisE[7];
        const user = new User({
          c: Date.now(), // should only have parenthesis upon creation not schema
          active: thisE[0],
          uid: thisE[1],
          username: thisE[2],
          email: thisE[3],
          pln: thisE[4],
          showFtns: thisE[5],
          display: thisE[6],
          password: bcrypt.hashSync(passwordPlainText, 8),
          overview: thisE[8],
        });
        user.save((err) => {
          if (err) {
            console.log(`+User ${thisV} error`, err);
            return;
          }
          console.log(`+User ${thisV}`);
          Role.findOne({ rid: "admin" }, (err, role) => {
            if (err) {
              console.log("  * no role: ", err);
              return;
            }
            user.roles = [role._id];
            user.save(err => {
              if (err) {
                console.log("  *", err);
                return;
              }
              user.uid = user.user_id.toString();
              user.save(err => {
                if (err) {
                  console.log("  *", err);
                  return;
                }
                console.log(`  * saved admin as uid ${user.uid} with password ${passwordPlainText}`);
              });
            });
          });
        });
      }
    }
    else {
      error = err;
    }
  });
}

function initialCategories(uid) {
  // entries:
  // (active,cid,caption,sid,display,parent,uid,pln)
  // (true,"art", "Browse", "art" "gallery", null, "0", 7),
  // (true,"forum", "Discuss", "forum" "forum", null, "0", 7)
  let adminUid = uid;
  Category.estimatedDocumentCount((err, count) => {
    if (!err && count === 0) {
      //[active,cid,caption,   sid,display,     parent,uid,pln]
      const entries = [
        [true, "art", "Browse", "art", "gallery", null, adminUid, 7],
        [true,"forum","Discuss","forum","forum",  null, adminUid, 7],
      ];
      for (var index = 0; index < entries.length; index++) {
        let thisE = entries[index];
        let thisV = thisE[1]; // let forces scope to be code block so last value doesn't repeat
        new Category({
          active: thisE[0],
          cid: thisE[1],
          caption: thisE[2],
          sid: thisE[3],
          display: thisE[4],
          parent: thisE[5],
          uid: thisE[6],
          pln: thisE[7],
        }).save((err) => {
          if (err) {
            console.log(`+Category ${thisV} by ${adminUid} (generated) error`, err);
          }
          else {
            console.log(`+Category ${thisV} by ${adminUid} (generated)`);
          }
        });
      }
    }
  });
}

function initialVoteTypes() {
  // entries:

  VoteType.estimatedDocumentCount((err, count) => {
    if (!err && count === 0) {
      // [active,vtn,v,caption,pln]
      const entries = [
        [true, 1, 1, "Like", 7],
        [true, 2, -1, "Dislike", 7],
        [true, 3, -1, "Offensive", 7],
        [true, 4, -1, "Spam", 7],
      ];
      for (var index = 0; index < entries.length; index++) {
        var thisE = entries[index];
        let thisV = thisE[1]; // let forces scope to be code block so last value doesn't repeat
        new VoteType({
          active: thisE[0],
          vtn: thisE[1],
          v: thisE[2],
          caption: thisE[3],
          pln: thisE[4],
        }).save((err) => {
          if (err) {
            console.log(`+VoteType ${thisV} error`, err);
          }
          else {
            console.log(`+VoteType ${thisV}`);
          }
        });
      }
    }
  });
}

function initialTestPosts(uid) {
  const adminUid = uid;
  Post.estimatedDocumentCount((err, count) => {
    if (!err && count === 0) {
      //[active,    uid,   title,   cid,  ftn, pln,         thumb, realRelPath, body]
      const entries = [
        [true, adminUid, "TestArt","art",  0,   7, "favicon.ico","favicon.ico", "This is an image description (generated by initialTestPost)."],
        [true, adminUid,"TestPost","forum",  0,   7,          null,         null, "This is the body of a forum post (generated by initialTestPost)."],
      ];
      for (var index = 0; index < entries.length; index++) {
        let thisE = entries[index];
        let thisV = thisE[2]; // let forces scope to be code block so last value doesn't repeat
        const post = new Post({
          c: Date.now(), // should only have parenthesis upon creation not schema
          active: thisE[0],
          uid: thisE[1],
          title: thisE[2],
          cid: thisE[3],
          ftn: thisE[4],
          pln: thisE[5],
          thumb: thisE[6],
          realRelPath: thisE[7],
          body: thisE[8],
        });
        post.save((err) => {
          if (err) {
            console.log(`+Post ${thisV} by ${adminUid} (generated) error`, err);
            return;
          }
          console.log(`+Post ${thisV} by ${adminUid} (generated)`);
          post.uid = adminUid;
          post.save(err => {
            if (err) {
              console.log("  *", err);
              return;
            }
            console.log(`  * saved a post "${post.title}" in ${post.cid}`);
          });
        });
      }
    }
  });
}
